### DISCUSSION TITLE
Parse generic numbers (int, hex, real)

### DISCUSSION THREAD

**Discussion by facontidavide** [foonathan/lexy#89] [Q&A] [answered]
Follow up of #85 

> The matching doesn't seem to work as expected if I do this: https://godbolt.org/z/jWG4zoKTP
> 
> Choice is ordered and takes the first branch that matches. When it sees 5.3 dsl::integer matches so that is parsed; it then leaves .3 unconsumed. I recommend attempting to match the real number first and ensuring that it does not match integers only: https://godbolt.org/z/dn83sxqdG

For the sake of completeness, this is the final solution I found to address negative integers and hexadecimals as well:

https://godbolt.org/z/Td76TsrKK

The only case that I am not handling correctly is the string `0x100.2` that should fail.

**Comment by foonathan** [comment, marked as answer]
> The only case that I am not handling correctly is the string 0x100.2 that should fail.

I mean, it does fail. It leaves a `.2` on the input which will cause subsequent parse operations to fail or exit early. If you add a `+ dsl::eof` to the root production, lexy will complain about the trailing input: https://godbolt.org/z/Y8dafdqzn

You can use `dsl::peek_not` to get a custom error if it happens: https://godbolt.org/z/cqzjdxbTx

You can even get a bit more fancy by recovering after the error: https://godbolt.org/z/nEbfd94oE

**Comment by facontidavide** [comment]
thanks a lot! 

Something that confuse me a little is this 
```c++
  static constexpr auto rule = [] {
    auto hex_integer = LEXY_LIT("0x") >> dsl::integer<int, dsl::hex>;
    auto regular_integer = dsl::peek(dsl::lit_c<'-'> / dsl::lit_c<'+'> / dsl::digit<>) >> dsl::p<integer>;
    auto float_error = dsl::peek_not(dsl::period / dsl::lit_c<'e'> / dsl::lit_c<'E'>).error<invalid_float_suffix>;
    return ( hex_integer | regular_integer ) >> float_error;
  }();
```
I expected it to be 

```c++
    return ( hex_integer | regular_integer ) + float_error;
```
That doesn't compile ("choice requires a branch condition").
Of course, there is something about the DSL that I am not grasping yet! :sweat_smile:

**Reply by foonathan** [reply]
You're using `dsl::p<Integer>` in a choice `dsl::p<Real> | dsl::p<Integer>`, which requires that both rules are branch rules. `dsl::p` is a branch rule if and only if the rule of the Production is a branch rule. `a + b` is not a branch rule, hence the error. `a >> b` is a branch rule, with `a` the condition. 

https://lexy.foonathan.net/learn/branching/
https://lexy.foonathan.net/reference/dsl/branch/#doc
https://lexy.foonathan.net/reference/dsl/choice/#choice

**Reply by facontidavide** [reply]
thanks!

**Comment by foonathan** [comment]
BTW, as of 0370d942e5bbae085c62c548e0244a2b052626a0 the recovery behavior of `dsl::peek_not` has been changed for this use case.
The following code checks if it has one or more characters that make up a float suffix, and if so, consumes them and issues an error covering them: https://godbolt.org/z/YbWxhE1dv

**Reply by facontidavide** [reply]
thanks. I will update my parser accordingly
